File: architecture/data/src/main/java/com/pacific/arch/data/IOError.java
package com.pacific.data;

import android.annotation.TargetApi;

public final class IOError extends RuntimeException {
    public final int code;

    public IOError(int code) {
        this.code = code;
    }

    public IOError(String message, int code) {
        super(message);
        this.code = code;
    }

    public IOError(String message, Throwable cause, int code) {
        super(message, cause);
        this.code = code;
    }

    public IOError(Throwable cause, int code) {
        super(cause);
        this.code = code;
    }

    @TargetApi(24)
    public IOError(String message, Throwable cause, boolean enableSuppression, boolean writableStackTrace, int code) {
        super(message, cause, enableSuppression, writableStackTrace);
        this.code = code;
    }

    public static boolean isIOError(Throwable throwable) {
        return throwable instanceof IOError;
    }

    public static IOError from(Throwable throwable) {
        if (isIOError(throwable)) {
            return (IOError) throwable;
        }
        return new IOError(throwable.getMessage(), throwable.getCause(), -1);
    }
}

File: architecture/arch/src/main/java/com/pacific/arch/data/IOError.kt
package com.pacific.arch.data

import android.annotation.TargetApi
import java.lang.RuntimeException

class IOError : RuntimeException {
    @JvmField
    val code: Int

    constructor(code: Int) : super() {
        this.code = code
    }

    constructor(message: String?, code: Int) : super(message) {
        this.code = code
    }

    constructor(cause: Throwable?, code: Int) : super(cause) {
        this.code = code
    }

    constructor(message: String?, cause: Throwable?, code: Int) : super(message, cause) {
        this.code = code
    }

    @TargetApi(24)
    constructor(message: String?,
                cause: Throwable?,
                enableSuppression: Boolean,
                writableStackTrace: Boolean,
                code: Int) : super(message, cause, enableSuppression, writableStackTrace) {
        this.code = code
    }

    companion object {
        fun isIOError(throwable: Throwable): Boolean {
            return throwable is IOError
        }

        fun from(throwable: Throwable): IOError {
            return throwable as? IOError ?: IOError(throwable.message!!, throwable.cause!!, -1)
        }
    }
}